File: frontend/src/lib/publishing.ts
/** Types + REST calls for publishing (Instagram). */
import { api } from "./api";

export interface SocialAccount {
  id: string;
  platform: string;
  account_name: string;
  ig_user_id: string | null;
  connected: boolean;
  is_dev: boolean;
}

export interface Post {
  id: string;
  social_account_id: string | null;
  trip_id: string | null;
  image_url: string | null;
  caption: string | null;
  status: "scheduled" | "publishing" | "published" | "failed";
  scheduled_at: string | null;
  published_at: string | null;
  result: { permalink?: string; mock?: boolean; detail?: string };
  created_at: string;
}

export const listAccounts = () => api<SocialAccount[]>("/api/v1/publishing/accounts");
export const connectDev = (account_name: string) =>
  api<SocialAccount>("/api/v1/publishing/connect/dev", {
    method: "POST",
    body: JSON.stringify({ account_name }),
  });
export const listPosts = () => api<Post[]>("/api/v1/publishing/posts");
export const createPost = (data: {
  caption: string;
  image_url?: string;
  creative_asset_id?: string;
  trip_id?: string;
}) => api<Post>("/api/v1/publishing/posts", { method: "POST", body: JSON.stringify(data) });
export const publishNow = (id: string) =>
  api<Post>(`/api/v1/publishing/posts/${id}/publish`, { method: "POST" });

File: backend/app/models/publishing.py
"""Publishing models (Module 4): social accounts and scheduled posts.

Tenant-scoped (RLS). Access tokens are stored encrypted (Fernet).
"""
import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TenantScoped, TimestampMixin, UUIDPrimaryKey

POST_STATUSES = ("scheduled", "publishing", "published", "failed")


class SocialAccount(Base, UUIDPrimaryKey, TimestampMixin):
    __tablename__ = "social_accounts"

    tenant_id: Mapped[uuid.UUID] = TenantScoped.tenant_fk()
    platform: Mapped[str] = mapped_column(String(20), nullable=False, default="instagram")
    account_name: Mapped[str] = mapped_column(String(160), nullable=False)
    ig_user_id: Mapped[str | None] = mapped_column(String(80), nullable=True)
    access_token_enc: Mapped[str | None] = mapped_column(Text, nullable=True)
    connected: Mapped[bool] = mapped_column(Boolean, default=True)
    is_dev: Mapped[bool] = mapped_column(Boolean, default=False)
    meta: Mapped[dict[str, Any]] = mapped_column(JSONB, default=dict)


class ScheduledPost(Base, UUIDPrimaryKey, TimestampMixin):
    __tablename__ = "scheduled_posts"

    tenant_id: Mapped[uuid.UUID] = TenantScoped.tenant_fk()
    social_account_id: Mapped[uuid.UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("social_accounts.id", ondelete="SET NULL"), nullable=True
    )
    trip_id: Mapped[uuid.UUID | None] = mapped_column(PGUUID(as_uuid=True), nullable=True)
    creative_asset_id: Mapped[uuid.UUID | None] = mapped_column(PGUUID(as_uuid=True), nullable=True)
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    caption: Mapped[str | None] = mapped_column(Text, nullable=True)
    scheduled_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    status: Mapped[str] = mapped_column(String(12), nullable=False, default="scheduled", index=True)
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    result: Mapped[dict[str, Any]] = mapped_column(JSONB, default=dict)
